File: app/src/main/java/com/example/android/simpledice/ui/HistoryActivity.java
package com.example.android.simpledice.ui;

import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.view.MenuCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.widget.TextView;

import com.example.android.simpledice.utils.Constants;
import com.example.android.simpledice.utils.DiceResults;
import com.example.android.simpledice.R;
import com.firebase.ui.auth.AuthUI;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdView;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import java.util.ArrayList;
import java.util.List;

public class HistoryActivity extends AppCompatActivity implements HistoryResultsAdapter.HistoryResultsClickHandler {

    private TextView mResultsNameTextView;
    private TextView mResultsTotalTextView;
    private TextView mResultsDescripTextView;
    private RecyclerView mHistoryRecyclerView;
    private HistoryResultsAdapter mHistoryAdapter;
    private AdView mAdView;

    private List<DiceResults> mDiceResults;

    //Firebase
    private String mUserID;
    private FirebaseAuth mFirebaseAuth;
    private FirebaseAuth.AuthStateListener mAuthStateListener;
    private FirebaseDatabase mFirebaseDatabase;
    private DatabaseReference mBaseDatabaseReference;
    private ChildEventListener mHistoryChildEventListener;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_history);

        setTitle(R.string.history_activity_title);

        initializeFirebase();

        assignViews();

        setupAds();

        setupHistoryRecyclerView();

    }

    @Override
    protected void onResume() {
        super.onResume();
        mFirebaseAuth.addAuthStateListener(mAuthStateListener);
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (mAuthStateListener != null) {
            mFirebaseAuth.removeAuthStateListener(mAuthStateListener);
        }
        detachDatabaseHistoryReadListener();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.history_activity_menu, menu);
        MenuCompat.setGroupDividerEnabled(menu, true);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case (R.id.action_sign_out):
                AlertDialog.Builder signoutBuilder = new AlertDialog.Builder(HistoryActivity.this);
                signoutBuilder.setTitle(R.string.sign_out_dialog_title)
                        .setMessage(R.string.sign_out_dialog_message)
                        .setPositiveButton(R.string.sign_out_dialog_positive, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                AuthUI.getInstance().signOut(HistoryActivity.this);
                            }
                        })
                        .setNegativeButton(R.string.sign_out_dialog_negative, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                dialog.cancel();
                            }
                        });
                signoutBuilder.show();
                return true;
            case (R.id.action_delete_all_history):
                AlertDialog.Builder deleteHistoryBuilder = new AlertDialog.Builder(HistoryActivity.this);
                deleteHistoryBuilder.setTitle(R.string.delete_all_dialog_title)
                        .setMessage(R.string.delete_all_dialog_message)
                        .setPositiveButton(R.string.delete_all_dialog_positive, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                clearAllHistory();
                            }
                        })
                        .setNegativeButton(R.string.delete_all_dialog_negative, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                dialog.cancel();
                            }
                        });
                deleteHistoryBuilder.show();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onItemClick(DiceResults historyResults) {
        setDataToResultsViews(historyResults);
    }

    /**
     * A helper method that assigns all of the views to their initial values in onCreate
     */
    private void assignViews() {
        mResultsNameTextView = findViewById(R.id.results_name_tv);
        mResultsTotalTextView = findViewById(R.id.results_total_tv);
        mResultsDescripTextView = findViewById(R.id.results_descrip_tv);
        mHistoryRecyclerView = findViewById(R.id.history_rv);
        mAdView = findViewById(R.id.banner_ad);
    }

    /**
     * A helper method to populate the results views with data
     *
     * @param diceResults to populate the views from
     */
    private void setDataToResultsViews(DiceResults diceResults) {
        mResultsNameTextView.setText(diceResults.getName());
        mResultsTotalTextView.setText(String.valueOf(diceResults.getTotal()));
        mResultsDescripTextView.setText(diceResults.getDescrip());
    }

    /**
     * Helper method to setup HistoryRecyclerView, should only be called once in onCreate
     */
    private void setupHistoryRecyclerView() {
        LinearLayoutManager layoutManager = new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false);
        mHistoryRecyclerView.setLayoutManager(layoutManager);

        mHistoryAdapter = new HistoryResultsAdapter(this);
        mHistoryRecyclerView.setAdapter(mHistoryAdapter);

        mHistoryRecyclerView.addItemDecoration(new DividerItemDecoration(this, DividerItemDecoration.VERTICAL));

        if (mDiceResults == null) {
            mDiceResults = new ArrayList<>();
        }
        mHistoryAdapter.setHistoryResults(mDiceResults);
    }

    /**
     * A helper method to handle all of the initial setup for Firebase, to be called in onCreate
     */
    private void initializeFirebase() {
        mUserID = Constants.INSTANCE.getFIREBASE_ANONYMOUS();
        //Auth setup
        mFirebaseAuth = FirebaseAuth.getInstance();

        //Database
        mFirebaseDatabase = FirebaseDatabase.getInstance();
        mBaseDatabaseReference = mFirebaseDatabase.getReference();

        mAuthStateListener = new FirebaseAuth.AuthStateListener() {
            @Override
            public void onAuthStateChanged(@NonNull FirebaseAuth firebaseAuth) {
                FirebaseUser user = firebaseAuth.getCurrentUser();
                if (user != null) { //Signed in
                    onSignedInInitialize(user.getUid());
                } else { //Signed out, finish activity
                    onSignedOutCleanup();
                    finish();
                }
            }
        };
    }

    /**
     * A helper method for when signed into FirebaseAuth
     *
     * @param userID to set the Activity's member variable
     */
    private void onSignedInInitialize(String userID) {
        mUserID = userID;
        attachDatabaseHistoryReadListener();
    }

    /**
     * A helper method for when signed out of FirebaseAuth
     */
    private void onSignedOutCleanup() {
        mUserID = Constants.INSTANCE.getFIREBASE_ANONYMOUS();
        detachDatabaseHistoryReadListener();
    }

    /**
     * A helper method for creating the database listener that checks Firebase for DiceResults objects //TODO change to single value event listener, add onSavedInstance?
     */
    private void attachDatabaseHistoryReadListener() {
        if (mHistoryChildEventListener == null) {
            mHistoryChildEventListener = new ChildEventListener() {
                @Override
                public void onChildAdded(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {
                    DiceResults diceResults = dataSnapshot.getValue(DiceResults.class);
                    mDiceResults.add(diceResults);
                    mHistoryAdapter.setHistoryResults(mDiceResults);
                }

                @Override
                public void onChildChanged(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {

                }

                @Override
                public void onChildRemoved(@NonNull DataSnapshot dataSnapshot) {

                }

                @Override
                public void onChildMoved(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {

                }

                @Override
                public void onCancelled(@NonNull DatabaseError databaseError) {

                }
            };
            mBaseDatabaseReference.child(Constants.INSTANCE.getFIREBASE_DATABASE_HISTORY_PATH()).child(mUserID).addChildEventListener(mHistoryChildEventListener);
        }
    }

    /**
     * A helper method to clear the database listener
     */
    private void detachDatabaseHistoryReadListener() {
        if (mHistoryChildEventListener != null) {
            mBaseDatabaseReference.child(Constants.INSTANCE.getFIREBASE_DATABASE_HISTORY_PATH()).child(mUserID).removeEventListener(mHistoryChildEventListener);
            mHistoryChildEventListener = null;
        }
    }

    /**
     * A helper method that deletes all the user's history from Firebase, and empties all views on HistoryActivity
     */
    private void clearAllHistory(){
        mBaseDatabaseReference.child(Constants.INSTANCE.getFIREBASE_DATABASE_HISTORY_PATH()).child(mUserID).setValue(null); //Delete from database

        //Reset stored values and update RecyclerView to empty
        mDiceResults = new ArrayList<>();
        mHistoryAdapter.setHistoryResults(mDiceResults);

        //Empty TextViews
        mResultsNameTextView.setText("");
        mResultsDescripTextView.setText("");
        mResultsTotalTextView.setText("");
    }

    /**
     * A helper method to setup an ad into the activity's AdView
     */
    private void setupAds() {
        AdRequest adRequest = new AdRequest.Builder().build();
        mAdView.loadAd(adRequest);
    }
}

File: app/src/main/java/com/example/android/simpledice/ui/HistoryActivity.kt
package com.example.android.simpledice.ui

import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.MenuCompat
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.android.simpledice.R
import com.example.android.simpledice.database.AllDiceResultsViewModel
import com.example.android.simpledice.database.AppDatabase
import com.example.android.simpledice.database.AppExecutors
import com.example.android.simpledice.utils.DiceResults
import com.google.android.gms.ads.AdRequest
import kotlinx.android.synthetic.main.activity_history.*
import kotlinx.android.synthetic.main.banner_ad.*
import kotlinx.android.synthetic.main.results.*
import java.util.*

class HistoryActivity : AppCompatActivity(), HistoryResultsAdapter.HistoryResultsClickHandler {

    private var mHistoryAdapter: HistoryResultsAdapter? = null

    private var mDiceResults: MutableList<DiceResults>? = null

    private var mDatabase : AppDatabase? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_history)

        setTitle(R.string.history_activity_title)

        restoreSavedInstanceState(savedInstanceState)

        setupAds()

        setupHistoryRecyclerView()

        setupDatabaseAndViewModel()

    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.history_activity_menu, menu)
        MenuCompat.setGroupDividerEnabled(menu, true)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_delete_all_history -> {
                val deleteHistoryBuilder = AlertDialog.Builder(this@HistoryActivity)
                deleteHistoryBuilder.setTitle(R.string.delete_all_dialog_title)
                    .setMessage(R.string.delete_all_dialog_message)
                    .setPositiveButton(R.string.delete_all_dialog_positive) { dialog, which -> clearAllHistory() }
                    .setNegativeButton(R.string.delete_all_dialog_negative) { dialog, which -> dialog.cancel() }
                deleteHistoryBuilder.show()
                return true
            }
            R.id.action_settings -> {
                val settingsIntent = Intent(this@HistoryActivity, SettingsActivity::class.java)
                startActivity(settingsIntent)
                return true
            }
            android.R.id.home -> {
                onBackPressed()
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    override fun onItemClick(historyResults: DiceResults) {
        setDataToResultsViews(historyResults)
    }


    /**
     * A helper method to populate the results views with data
     *
     * @param diceResults to populate the views from
     */
    private fun setDataToResultsViews(diceResults: DiceResults) {
        results_name_tv.text = diceResults.name
        results_total_tv.text = diceResults.total.toString()
        results_descrip_tv.text = diceResults.descrip
    }

    /**
     * Helper method to setup HistoryRecyclerView, should only be called once in onCreate
     */
    private fun setupHistoryRecyclerView() {
        val layoutManager = LinearLayoutManager(
            this,
            RecyclerView.VERTICAL,
            false
        )
        history_rv.layoutManager = layoutManager

        mHistoryAdapter = HistoryResultsAdapter(this)
        history_rv.adapter = mHistoryAdapter

        history_rv.addItemDecoration(
            DividerItemDecoration(
                this,
                DividerItemDecoration.VERTICAL
            )
        )

        if (mDiceResults == null) {
            mDiceResults = ArrayList()
        }
        mHistoryAdapter!!.setHistoryResults(mDiceResults!!)
    }

    /**
     * A helper method that deletes all the user's history from Firebase, and empties all views on HistoryActivity
     */
    private fun clearAllHistory() {
        AppExecutors.getInstance()!!.diskIO.execute {
            mDatabase!!.diceResultsDao().deleteAllDiceResults()
        }

        //Reset stored values and update RecyclerView to empty
        mDiceResults = ArrayList()
        mHistoryAdapter!!.setHistoryResults(mDiceResults!!)

        //Empty TextViews
        results_name_tv.text = ""
        results_descrip_tv.text = ""
        results_total_tv.text = ""
    }

    /**
     * A helper method to setup an ad into the activity's AdView
     */
    private fun setupAds() {
        val adRequest = AdRequest.Builder().build()
        banner_ad.loadAd(adRequest)
    }

    /**
     * A helper method for preparing access to the database, and populating views relevant to it
     */
    private fun setupDatabaseAndViewModel() {
        mDatabase = AppDatabase.getInstance(this)

        val viewModel = ViewModelProviders.of(this).get(AllDiceResultsViewModel::class.java)
        viewModel.diceResults!!.observe(this, androidx.lifecycle.Observer { databaseDiceResults ->

            //Hide Recycler, show ProgressBar
            history_rv.visibility = View.GONE
            history_progress_bar.visibility = View.VISIBLE

            mDiceResults = arrayListOf() //Clear and repopulate mDiceRolls
            for (diceResults in databaseDiceResults) {
                mDiceResults!!.add(diceResults)
            }
            mHistoryAdapter!!.setHistoryResults(mDiceResults!!)

            //Show Recycler, hide ProgressBar
            history_rv.visibility = View.VISIBLE
            history_progress_bar.visibility = View.GONE
        })
    }

    /**
     * Override, save the values of the results view for later restoration
     */
    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)

        //Include the name, descrip, and total from the results views
        outState.putString(getString(R.string.history_saved_instance_name_key), results_name_tv.text.toString())
        outState.putString(getString(R.string.history_saved_instance_descrip_key), results_descrip_tv.text.toString())
        outState.putString(getString(R.string.history_saved_instance_total_key), results_total_tv.text.toString())
    }



    /**
     * A helper method that restores the savedInstanceState on rotation, called in onCreate
     * @param savedInstanceState to restore, provided by onCreate
     */
    private fun restoreSavedInstanceState(savedInstanceState: Bundle?){
        if (savedInstanceState != null){
            //Get saved values, restore to textviews (default to empty string if unavailable)
            val name = savedInstanceState.getString(getString(R.string.history_saved_instance_name_key), "")
            val total = savedInstanceState.getString(getString(R.string.history_saved_instance_total_key), "")
            val descrip = savedInstanceState.getString(getString(R.string.history_saved_instance_descrip_key), "")

            results_name_tv.text = name
            results_total_tv.text = total
            results_descrip_tv.text = descrip

        } else { //Provide a hint to click on recycler item if there is no savedInstanceState
            results_descrip_tv.text = getString(R.string.history_click_a_result)
        }
    }

    override fun onBackPressed() {
        //If over version 21, scroll the results to start to make activity transitions appear smoother
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            descrip_scrollview.smoothScrollTo(0, 0)
        }
        super.onBackPressed()
    }
}
